### PR TITLE
Added documentation for inlet outlet

### PR DESCRIPTION
**Pull request by pawansnegi**
Added documentation for inlet outlet
Small fix in crksph and fpc

### CODE REVIEW

File: docs/source/design/iom.rst [hunk 1]
```diff
@@ -0,0 +1,86 @@
+.. _iom:
+
+==================
+Writing Inlet oulet manager (IOM)
+==================
+
+This section essentially talks about writing your own inlet outlet manager.
```
Comment (prabhuramachandran) [inline docs/source/design/iom.rst:7]: Remove "essentially", use "discusses writing your own ..."

File: docs/source/design/iom.rst [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        def add_io_properties(self, pa, scheme=None):
+            # Add properties to be used in inlet/outlet equations tp pa
+            # pass
+
+        def get_io_names(self, ghost=False):
+            # return all the names of inlets and outlets
+            
+        def get_stepper(self, scheme, integrator, **kw):
+            # Returns the steppers for inlet/outlet
+            raise NotImplementedError()
```
Comment (prabhuramachandran) [inline docs/source/design/iom.rst:45]: Remove this.

File: docs/source/design/iom.rst [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    class MyIOM(InletOutletManager):
+        def __init__(self, fluid_arrays, inletinfo, outletinfo,
+                    extraeqns=None):
+            # Create the object to manage inlet outlet boundary conditions.
+            # Most of the variables are evaluated after the scheme and particles
+            # are created.
+
+        def create_ghost(self, pa_arr, inlet=True):
+            # Creates ghosts for the given inlet/outlet particles
+            # return ghost_pa ( the ghost particle array for the pa_arr)
```
Comment (prabhuramachandran) [inline docs/source/design/iom.rst:31]: Can one return an empty list if there are no ghosts needed?

File: docs/source/design/iom.rst [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        def get_io_names(self, ghost=False):
+            # return all the names of inlets and outlets
+            
+        def get_stepper(self, scheme, integrator, **kw):
+            # Returns the steppers for inlet/outlet
+            raise NotImplementedError()
+
+        def setup_iom(self, dim, kernel):
+            # Essential data passed
```
Comment (prabhuramachandran) [inline docs/source/design/iom.rst:48]: Not clear what this essential data is?

Comment (prabhuramachandran) [reply docs/source/design/iom.rst:48]: Why isn't this in the `__init__`?

File: docs/source/design/iom.rst [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+- The ``get_equations`` and ``get_equations_post_compute_acceleration`` provides the additional equations to be used to interpolate properties from fluid particle arrays. This is to be called in ``create_equations`` method of the :py:class:`Scheme` instance. 
+
+- The ``get_inlet_outlet`` methods provides the instances for the :py:class:`InletInfo` and :py:class:`OutletInfo` which updates the particles when they cross the interface. This method is called in ``create_inlet_outlet`` method of the :py:class:`Application` instance. 
+
+- Any additional data required from the :py:class:`Application` instance can be passed using ``setup_iom`` method. 
+
+- In mirror type inlet-outlet a ghost layer of particles is required which is a mere reflection about the inlet/outlet-fluid interface. It is created in ``create_particles`` using ``create_ghost``. 
+
+The IOM enables the management of the above steps easy to handle.
```
Comment (prabhuramachandran) [inline docs/source/design/iom.rst:86]: You can again add a reference to the example you cited above.

File: docs/source/examples/flow_past_cylinder.rst [hunk 6]
```diff
@@ -0,0 +1,252 @@
+.. _flow_past_cylinder:
+
+Flow past a circular cylinder using open boundary conditions
+----------------------------------------------
+
+This example demonstrates the API of inlet and outlet boundary conditions in PySPH. The flow past a circular cylinder is one the example which uses both inlet and outlet boundary condition.
```
Comment (prabhuramachandran) [inline docs/source/examples/flow_past_cylinder.rst:6]: "is one the example" -> "is an example"

Comment (prabhuramachandran) [reply docs/source/examples/flow_past_cylinder.rst:6]: "inlet and outlet boundary conditions"

File: docs/source/examples/flow_past_cylinder.rst [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+To run it one may do::
+
+  $ cd ~/pysph/pysph/examples/
+  $ python flow_past_cylinder_2d.py
```
Comment (prabhuramachandran) [inline docs/source/examples/flow_past_cylinder.rst:10]: Use `pysph run flow_past_cylinder_2d`

File: docs/source/examples/flow_past_cylinder.rst [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+There are many command line options that this example provides, check them out with::
+
+  $ python flow_past_cylinder_2d.py -h
+
+The example source can be seen at `flow_past_cylinder_2d.py
+<https://github.com/pypr/pysph/tree/master/pysph/examples/flow_past_cylinder_2d.py>`_.
+
+
```
Comment (prabhuramachandran) [inline docs/source/examples/flow_past_cylinder.rst:19]: Maybe describe the problem and the boundary conditions here so it is easier to understand?

File: docs/source/examples/flow_past_cylinder.rst [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+Creating ghost particles
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this example, we have a wind tunnel with two bounding slip walls on the
+top and below of the tunnel. The inlet is on the left and the outlet is on
+the right. In order to perform the simulation five particle arrays,
+``solid``, ``fluid``, ``wall``, ``inlet`` and ``outlet`` are generated.
+``fluid``, ``solid`` and ``wall`` has to obey ``edac`` scheme, whereas
+``inlet`` and ``outlet`` are solved according to the boundary conditions.
```
Comment (prabhuramachandran) [inline docs/source/examples/flow_past_cylinder.rst:118]: This is a good starting paragraph, maybe it should start with this and then go into IOM details?

File: docs/source/examples/flow_past_cylinder.rst [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+the IOM depending upon the type of IOM subclass used. The properties
+:math:`uag`, :math:`uta` are the time average and velocity array in :math:`x`
+direction at t=0.
+
+Creating updater
+~~~~~~~~~~~~~~~~~
+
+The purpose of the updater is to remove particle from ``inlet`` and add them
+to ``fluid`` whenever a particle crosses the inlet-outlet interface. Similarly,
+it is done in case of the ``oulet``. It also adds new particle to ``inlet`` as required and remove a particle from the simulation when they flow past ``outlet``.
```
Comment (prabhuramachandran) [inline docs/source/examples/flow_past_cylinder.rst:158]: Is this done automatically or does the user need to write anything extra?

File: docs/source/examples/flow_past_cylinder.rst [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+
+
+2. The IOM must be configured in the ``configure_scheme`` function.
+
+3. In case you change the integrator of the function, make sure the updater
+``io`` is updating in the appropriate stage. For example, in case of a
+``PECIntegrator`` class of integrator, the particles integrated half step in
+stage 1 and finally advected in stage 2 then ``io`` updates the particle
+arrays after stage 2 is complete. In case one wants to do the update in stage
+1 (while using another integrator) the arguments must be passed to the updater appropriately.
```
Comment (prabhuramachandran) [inline docs/source/examples/flow_past_cylinder.rst:201]: How does one set the properties of the inlet/outlet?  Are those done with equations?  Is there an example that users should read?

**Review by prabhuramachandran — CHANGES_REQUESTED** [review]
Looks great, please look at the suggestions.  Thanks.

File: docs/source/design/iom.rst (hunk 1 above)
Comment (pawansnegi) [reply docs/source/design/iom.rst:7]: Done

File: docs/source/design/iom.rst (hunk 3 above)
Comment (pawansnegi) [reply docs/source/design/iom.rst:31]: Yes

File: docs/source/design/iom.rst (hunk 4 above)
Comment (pawansnegi) [reply docs/source/design/iom.rst:48]: It is required to pass information from the consume_user_data. Added this in the documentation.

File: docs/source/design/iom.rst (hunk 5 above)
Comment (pawansnegi) [reply docs/source/design/iom.rst:86]: done

File: docs/source/examples/flow_past_cylinder.rst (hunk 6 above)
Comment (pawansnegi) [reply docs/source/examples/flow_past_cylinder.rst:6]: done

File: docs/source/examples/flow_past_cylinder.rst (hunk 7 above)
Comment (pawansnegi) [reply docs/source/examples/flow_past_cylinder.rst:10]: done

File: docs/source/examples/flow_past_cylinder.rst (hunk 8 above)
Comment (pawansnegi) [reply docs/source/examples/flow_past_cylinder.rst:19]: done

File: docs/source/examples/flow_past_cylinder.rst (hunk 9 above)
Comment (pawansnegi) [reply docs/source/examples/flow_past_cylinder.rst:118]: done

File: docs/source/examples/flow_past_cylinder.rst (hunk 10 above)
Comment (pawansnegi) [reply docs/source/examples/flow_past_cylinder.rst:158]: yes done automatically. Added that in the doc.

File: docs/source/examples/flow_past_cylinder.rst (hunk 11 above)
Comment (pawansnegi) [reply docs/source/examples/flow_past_cylinder.rst:201]: Added that in the Writing IOM section.

**Review by prabhuramachandran — APPROVED** [review]

**Pull request merged**